File: Assets/src/model/Tile.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

[Serializable]
public abstract class Tile : Entity {
  public TileVisiblity visibility = TileVisiblity.Unexplored;
  private Vector2Int _pos;
  public bool isExplored => !IsDead && tile.visibility != TileVisiblity.Unexplored;

  public override Vector2Int pos {
    get => _pos;
    /// do not allow moving tiles
    set { }
  }

  public override IEnumerable<object> MyModifiers => base.MyModifiers.Append(grass).Append(item);

  public Tile(Vector2Int pos) : base() {
    this._pos = pos;
  }

  protected override void HandleEnterFloor() {
    /// set correct visibility when the tile is dynamically added
    var player = GameModel.main?.player;
    if (player != null) {
      if (floor == player.floor && floor.TestVisibility(player.pos, pos)) {
        visibility = TileVisiblity.Visible;
      }
    }
  }

  /// 0.0 means unwalkable.
  /// weight 1 is "normal" weight.
  public float GetPathfindingWeight() => body != null ? 0 : BasePathfindingWeight();

  internal void BodyLeft(Body body) {
    if (body is Actor actor) {
      GameModel.main.EnqueueEvent(() => {
        foreach (var handler in this.Of<IActorLeaveHandler>()) {
          handler.HandleActorLeave(actor);
        }
      });
    }
  }

  internal void BodyEntered(Body body) {
    if (body is Actor actor) {
      GameModel.main.EnqueueEvent(() => {
        foreach (var handler in this.Of<IActorEnterHandler>()) {
          handler.HandleActorEnter(actor);
        }
      });
    }
  }

  public virtual float BasePathfindingWeight() {
    return 1;
  }

  public virtual bool ObstructsVision() {
    return BasePathfindingWeight() == 0 || (body is IBlocksVision || grass is IBlocksVision);
  }

  internal bool CanBeOccupied() {
    return GetPathfindingWeight() != 0;
  }
}

public interface IActorEnterHandler {
  void HandleActorEnter(Actor who);
}

public interface IActorLeaveHandler {
  void HandleActorLeave(Actor who);
}

[Serializable]
public enum TileVisiblity {
  Unexplored, Visible, Explored
}

[Serializable]
[ObjectInfo(description: "Creatures can walk here. Grass can grow here.", flavorText: "Earth - the first element.")]
public class Ground : Tile {
  public Ground(Vector2Int pos) : base(pos) { }
}

[Serializable]
[ObjectInfo("sign")]
public class Signpost : Ground {
  public bool hasRead;
  public string text;
  public Signpost(Vector2Int pos, string text = "") : base(pos) {
    this.text = text;
    hasRead = PlayerPrefs.HasKey("read-tip-"+text.GetHashCode());
  }

  public void ShowSignpost() {
    Popups.CreateStandard(
      title: "Tip",
      category: "",
      /// hack - add a linebreak before to put some space
      info: "\n" + text,
      flavor: "",
      sprite: null
    );
    hasRead = true;
    PlayerPrefs.SetInt("read-tip-"+text.GetHashCode(), 1);
  }
}

[Serializable]
[ObjectInfo(description: "Grass cannot grow on Hard Ground.", flavorText: "Any workable earth has been blown or washed away.")]
public class HardGround : Tile {
  public HardGround(Vector2Int pos) : base(pos) { }

  protected override void HandleEnterFloor() {
    base.HandleEnterFloor();
    grass?.Kill(this);
  }
}

[Serializable]
[ObjectInfo(flavorText: "")]
public class FancyGround : Ground {
  public FancyGround(Vector2Int pos) : base(pos) {
  }
}

[ObjectInfo(description: "Blocks vision and movement.", flavorText: "Hard earth that has weathered centuries of erosion; it's not going anywhere.")]
[Serializable]
public class Wall : Tile {
  public Wall(Vector2Int pos) : base(pos) { }
  public override float BasePathfindingWeight() {
    return 0;
  }
}

[Serializable]
[ObjectInfo(description: "Blocks movement.", flavorText: "You look down and cannot see the bottom. Be careful not to fall!")]
public class Chasm : Tile {
  public Chasm(Vector2Int pos) : base(pos) {
  }

  protected override void HandleEnterFloor() {
    base.HandleEnterFloor();
    // remove bodies and grasses on it. don't "kill" it though, since technically they're not dead
    if (grass != null) {
      floor.Remove(grass);
    }
    if (body != null && !(body is Actor)) {
      floor.Remove(body);
    }
    if (item != null) {
      floor.Remove(item);
      // this will trigger the item placement behavior and it will find an acceptable spot
      floor.Put(item);
    }
  }

  public override float BasePathfindingWeight() {
    return 0;
  }

  public override bool ObstructsVision() {
    return body is IBlocksVision || grass is IBlocksVision;
  }
}

interface IAlwaysVisibleTile {}

[Serializable]
[ObjectInfo(description: "Go back home.")]
public class Upstairs : Tile, IOnTopActionHandler, IAlwaysVisibleTile {
  /// <summary>Where the player will be after taking the Downstairs connected to this tile.</summary>
  public Vector2Int landing => pos + Vector2Int.right;

  public string OnTopActionName => "Go Home";

  public void HandleOnTopAction() {
    GameModel.main.player.task = new GenericPlayerTask(
      GameModel.main.player,
      TryGoHome
    );
  }

  public Upstairs(Vector2Int pos) : base(pos) {
    visibility = TileVisiblity.Visible;
   }

  protected override void HandleEnterFloor() {
    base.HandleEnterFloor();
    grass?.Kill(this);
  }

  public void TryGoHome() {
    if (floor.EnemiesLeft() == 0) {
      Serializer.SaveMainToCheckpoint();
      GameModel.main.PutPlayerAt(0);
    }
  }
}

[Serializable]
[ObjectInfo(description: "Go deeper into the dungeon.")]
public class Downstairs : Tile, IOnTopActionHandler, IAlwaysVisibleTile {
  /// <summary>Where the player will be after taking the Upstairs connected to this tile.</summary>
  public Vector2Int landing => pos + Vector2Int.left;

  public string OnTopActionName => "Descend";
  public void HandleOnTopAction() {
    GameModel.main.player.task = new GenericPlayerTask(
      GameModel.main.player,
      TryGoDownstairs
    );
  }

  public Downstairs(Vector2Int pos) : base(pos) {
    visibility = TileVisiblity.Visible;
  }

  protected override void HandleEnterFloor() {
    base.HandleEnterFloor();
    grass?.Kill(this);
  }

  public void TryGoDownstairs() {
    floor.PlayerGoDownstairs();
  }
}

[ObjectInfo(description: "Plant seeds here.", flavorText: "Fresh, moist, and perfect for growing. Hard to come by in the caves.")]
[Serializable]
public class Soil : Tile {
  public Soil(Vector2Int pos) : base(pos) { }
}

[ObjectInfo("water_0", description: "Walk into to collect.", flavorText: "Water water everywhere...")]
[Serializable]
public class Water : Tile, IActorEnterHandler {
  public Water(Vector2Int pos) : base(pos) {
  }

  public void Collect() {
    GameModel.main.player.water += MyRandom.Range(55, 65);
    floor.Put(new Ground(pos));
  }

  public void HandleActorEnter(Actor who) {
    if (who == GameModel.main.player) {
      Collect();
    }
  }
}

File: web/src/model/Tile.ts
import { Entity } from './Entity';
import { Vector2Int } from '../core/Vector2Int';
import { TileVisibility, CollisionLayer, type IBlocksMovement, type IActorEnterHandler, type IActorLeaveHandler, ACTOR_ENTER_HANDLER, ACTOR_LEAVE_HANDLER } from '../core/types';
import { collectModifiers } from '../core/Modifiers';
import { GameModelRef } from './GameModelRef';

/** Implement on Body or Grass to block line-of-sight */
export interface IBlocksVision {
  readonly blocksVision: true;
}

/**
 * Abstract base for all tile types.
 * Port of C# Tile.cs.
 */
export abstract class Tile extends Entity {
  visibility: TileVisibility = TileVisibility.Unexplored;
  private _pos: Vector2Int;

  get pos(): Vector2Int {
    return this._pos;
  }

  /** Tiles don't move */
  set pos(_value: Vector2Int) {}

  get myModifiers(): Iterable<object | null | undefined> {
    return [...super.myModifiers, this.grass, this.item];
  }

  constructor(pos: Vector2Int) {
    super();
    this._pos = pos;
  }

  protected handleEnterFloor(): void {
    const player = GameModelRef.mainOrNull?.player;
    if (player && this.floor === player.floor) {
      if (this.floor!.testVisibility(player.pos, this.pos)) {
        this.visibility = TileVisibility.Visible;
      }
    }
  }

  /** Which collision layers this tile blocks. */
  get blocksMovement(): CollisionLayer {
    return this.basePathfindingWeight() === 0
      ? CollisionLayer.All
      : CollisionLayer.None;
  }

  /** 0 = unwalkable, 1 = normal weight */
  getPathfindingWeight(): number {
    return this.body != null ? 0 : this.basePathfindingWeight();
  }

  /** Pathfinding weight considering a specific body's movement layers */
  getPathfindingWeightFor(mover: any): number {
    if (this.body != null) return 0;
    let blocked = this.blocksMovement;
    const g = this.grass;
    if (g && 'blockedLayers' in g) {
      blocked |= (g as IBlocksMovement).blockedLayers;
    }
    return (mover.movementLayer & ~blocked) !== CollisionLayer.None ? 1 : 0;
  }

  bodyLeft(body: any): void {
    if (body.faction !== undefined && GameModelRef.mainOrNull) {
      const actor = body;
      GameModelRef.main.enqueuEvent(() => {
        for (const handler of collectModifiers<IActorLeaveHandler>(this, ACTOR_LEAVE_HANDLER)) {
          handler.handleActorLeave(actor);
        }
      });
    }
  }

  bodyEntered(body: any): void {
    if (body.faction !== undefined && GameModelRef.mainOrNull) {
      const actor = body;
      GameModelRef.main.enqueuEvent(() => {
        for (const handler of collectModifiers<IActorEnterHandler>(this, ACTOR_ENTER_HANDLER)) {
          handler.handleActorEnter(actor);
        }
      });
    }
  }

  basePathfindingWeight(): number {
    return 1;
  }

  obstructsVision(): boolean {
    if (this.basePathfindingWeight() === 0) return true;
    const b = this.body;
    const g = this.grass;
    return (b && 'blocksVision' in b) || (g && 'blocksVision' in g) || false;
  }

  canBeOccupied(): boolean {
    if (this.body != null) return false;
    let blocked = this.blocksMovement;
    const g = this.grass;
    if (g && 'blockedLayers' in g) {
      blocked |= (g as IBlocksMovement).blockedLayers;
    }
    return (blocked & CollisionLayer.Walking) === 0;
  }

  canBeOccupiedBy(mover: any): boolean {
    if (this.body != null) return false;
    let blocked = this.blocksMovement;
    const g = this.grass;
    if (g && 'blockedLayers' in g) {
      blocked |= (g as IBlocksMovement).blockedLayers;
    }
    return (mover.movementLayer & ~blocked) !== CollisionLayer.None;
  }
}

// --- Concrete tile types ---

export class Ground extends Tile {
  constructor(pos: Vector2Int) {
    super(pos);
  }
}

export class Signpost extends Ground {
  hasRead = false;
  text: string;

  constructor(pos: Vector2Int, text = '') {
    super(pos);
    this.text = text;
  }
}

export class HardGround extends Tile {
  constructor(pos: Vector2Int) {
    super(pos);
  }

  protected handleEnterFloor(): void {
    super.handleEnterFloor();
    this.grass?.killSelf();
  }
}

export class FancyGround extends Ground {
  constructor(pos: Vector2Int) {
    super(pos);
  }
}

export class Wall extends Tile {
  constructor(pos: Vector2Int) {
    super(pos);
  }

  get blocksMovement(): CollisionLayer {
    return CollisionLayer.All;
  }

  basePathfindingWeight(): number {
    return 0;
  }
}

export class Chasm extends Tile {
  constructor(pos: Vector2Int) {
    super(pos);
  }

  /** Chasms block walking but not flying */
  get blocksMovement(): CollisionLayer {
    return CollisionLayer.Walking;
  }

  protected handleEnterFloor(): void {
    super.handleEnterFloor();
    if (this.grass != null) {
      this.floor!.remove(this.grass);
    }
    if (this.item != null) {
      this.floor!.remove(this.item);
      this.floor!.put(this.item);
    }
  }

  basePathfindingWeight(): number {
    return 0;
  }

  obstructsVision(): boolean {
    const b = this.body;
    const g = this.grass;
    return (b && 'blocksVision' in b) || (g && 'blocksVision' in g) || false;
  }
}

export class Soil extends Tile {
  constructor(pos: Vector2Int) {
    super(pos);
  }
}

export class Water extends Tile implements IActorEnterHandler {
  readonly [ACTOR_ENTER_HANDLER] = true as const;
  private collected = false;

  constructor(pos: Vector2Int) {
    super(pos);
  }

  handleActorEnter(who: any): void {
    const player = GameModelRef.mainOrNull?.player;
    if (who === player) {
      this.collect();
    }
  }

  private collect(): void {
    this.collected = true;
    // In the daily puzzle version, water just becomes ground (no water meter)
    this.floor!.put(new Ground(this.pos));
  }
}
